fs/fs_zenfs.cc

```diff
@@ -1274,7 +1274,124 @@ void ZenFS::GetZenFSSnapshot(ZenFSSnapshot& snapshot,
+    if(ends_with(fname, ".sst")) {
+      file_extents[fname].emplace_back(ext);
+    }
+  }
+
+  for (const auto& it : file_extents) {
+    MigrateFileExtents(it.first, it.second, direct_io);
+  }
+}
+
+void ZenFS::MigrateFileExtents(
+    const std::string& fname,
+    const std::vector<ZoneExtentSnapshot*>& migrate_exts,
+    bool direct_io) {
+  IOStatus s;
+  Info(logger_, "MigrateFileExtents, fname: %s, extent count: %lu",
+       fname.data(), migrate_exts.size());
+  // The file may be deleted by other threads, better double check.
+  auto zfile = GetFile(fname);
+  if (zfile == nullptr || zfile->IsOpenForWR()) {
+    return;
+  }
+
+  std::vector<ZoneExtent*> extents = zfile->GetExtents();
+  for (ZoneExtent* ext : extents) {
+    // Check if current extent need to be migrated
+    auto it = std::find_if(migrate_exts.begin(), migrate_exts.end(),
+                           [&](const ZoneExtentSnapshot* ext_snapshot) {
+                             return ext_snapshot->start == ext->start_ &&
+                                    ext_snapshot->length == ext->length_;
+                           });
+
+    if (it == migrate_exts.end()) {
+      Info(logger_, "Migrate extent not found, ext_start: %lu", ext->start_);
+      continue;
+    }
+
+    Zone* old_zone = ext->zone_;
+    Zone* target_zone = nullptr;
+
+    // Allocate a new migration zone.
+    zbd_->TakeMigrateZone(&target_zone, zfile->GetWriteLifeTimeHint(),
+                          ext->length_);
+    if (target_zone == nullptr) {
+      zbd_->ReleaseMigrateZone(target_zone);
+      Info(logger_, "Migrate Zone Acquire Failed, Ignore Task.");
+      continue;
+    }
+
+    uint64_t target_start = target_zone->wp_;
+    if(!direct_io) {
+      // For buffered write, ZenFS use inlined metadata for extents and each
+      // extent has a SPARSE_HEADER_SIZE.
+      target_start = target_zone->wp_ + ZoneFile::SPARSE_HEADER_SIZE;
+      zfile->MigrateData(ext->start_ - ZoneFile::SPARSE_HEADER_SIZE,
+                         ext->length_ + ZoneFile::SPARSE_HEADER_SIZE,
+                         target_zone);
+    } else {
+      zfile->MigrateData(ext->start_, ext->length_, target_zone);
+    }
+
+    // Check again if the file still exist
+    if (GetFileInternal(fname) == nullptr) {
+      Info(logger_, "Migrate file not exist anymore.");
+      zbd_->ReleaseMigrateZone(target_zone);
+      break;
+    }
+
+#ifndef NDEBUG
+    // Double check migrated data
+    if (!zbd_->IsDataIdentical(ext->start_, target_start, ext->length_)) {
+      Info(logger_,
+           "Double Check Migration Data Failed, fname = %s, From %lu To %lu",
+           fname.data(), ext->start_, target_start);
+    } else {
+      Info(logger_,
+           "Double Check Migration Data Success, fname = %s, From %lu To %lu",
+           fname.data(), ext->start_, target_start);
+    }
+#endif
+
+    // Update zone stats
+    files_mtx_.lock();
+    ext->start_ = target_start;
+    ext->zone_ = target_zone;
```

Comment:
This is truly a problem, nice catch! If a thread is holding an old extent's offset to read, after we decrease old_zone's used_capacity, it could be reset. Let's figure out how to reset safely....

Reply:
To address this problem, I did the following changes:
1. Add a simple read-write lock inside ZoneExtent (see the code details)
2. Before we actually change the extent's start_ position, we should acquire a write lock and if another reader reading from the file, we cannot change extent's info
3. The critical section is quite small.

Do you have any better idea to replace the read-write lock here?

Reply:
Looking at the code, I think it would be better to add an exclusive per-file extent-list lock(in stead of a per-extent lock), that would be held during read(file read) / update(when migration is complete) operations.

Sorry for suggesting a per-extent lock, that looks just impractical without doing a lot of refactoring work. 

The migration flow would then be something like this:

1. Copy the file extents list
2. Move extents data and update the extents list copy
3. If the file has not been deleted
3.1 Write a replace metadata record with the new extents
3.2 Replace the file original extents with the cloned(migrated) extents, holding the extents lock
3.3 Deallocate the moved(old) extents

Reply:
Took your advice, change to per-file extent list lock

Reply:
Looks allright now